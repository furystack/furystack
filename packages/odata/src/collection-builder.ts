import { OdataCount } from './actions/count'
import { Collection } from './models/collection'

/**
 * Builder class for OData Entities
 */
export class CollectionBuilder {
  public readonly collections: Map<string, Collection<any>> = new Map()

  public addCollection<T>(collection: Collection<T>) {
    collection.functions = {
      ...collection.functions,
      $count: {
        action: OdataCount,
        isBound: true,
        returnType: Number,
      },
    }

    this.collections.set(collection.name || collection.model.name, {
      ...collection,
    })
    return this
  }
}
