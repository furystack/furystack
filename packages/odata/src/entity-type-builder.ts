import { Entity } from './models/entity'

/**
 * Builder class for OData Entities
 */
export class EntityTypeBuilder {
  public readonly entities: Map<string, Entity<any>> = new Map()

  public addEntityType<T>(entity: Entity<T>) {
    this.entities.set(entity.name || entity.model.name, entity)
    return this
  }
}
