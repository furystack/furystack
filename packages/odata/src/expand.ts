import { Injector } from '@furystack/inject'
import { Repository } from '@furystack/repository'
import { Entity, NavigationProperty, NavigationPropertyCollection } from './models'

/**
 * Method that adds expanded fields to an entity model
 * @param options The options to be provided
 */
export const expand = async <T>(options: {
  injector: Injector
  entity: T
  entityType: Entity<T>
  fieldNames?: Array<keyof T>
  repo: Repository
}) => {
  if (!options.fieldNames || options.fieldNames.length === 0) {
    return options.entity
  }

  const navProperties: Array<NavigationProperty<any> | NavigationPropertyCollection<any>> = options.fieldNames
    .map(
      fieldName =>
        (options.entityType.navigationProperties &&
          options.entityType.navigationProperties.find(np => np.propertyName === fieldName)) ||
        (undefined as any),
    )
    .filter(np => np !== undefined)

  if (!navProperties.length) {
    return options.entity
  }

  const expandedEntity = { ...options.entity } as T & { [key: string]: any }
  await Promise.all(
    navProperties.map(async navProperty => {
      const dataSet = options.repo.getDataSetFor(navProperty.dataSet || navProperty.relatedModel)
      if ((navProperty as NavigationProperty<any>).getRelatedEntity) {
        const expanded = await (navProperty as NavigationProperty<any>).getRelatedEntity(
          options.entity,
          dataSet,
          options.injector,
        )
        expandedEntity[navProperty.propertyName] = expanded
      } else if ((navProperty as NavigationPropertyCollection<any>).getRelatedEntities) {
        const expanded = await (navProperty as NavigationPropertyCollection<any>).getRelatedEntities(
          options.entity,
          dataSet,
          options.injector,
        )
        expandedEntity[navProperty.propertyName] = expanded
      }
    }),
  )
  return expandedEntity
}
