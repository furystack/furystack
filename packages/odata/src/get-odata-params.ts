import { parse } from 'url'
import { Entity, EdmType } from './models'

export type OrderType<T> = { [key in keyof T]?: 'ASC' | 'DESC' }
export interface OdataParams<T> {
  select: Array<keyof T>
  expand: Array<keyof T>
  orderBy: OrderType<T> | undefined
  expandExpression: string
  filter: { [key in keyof T]?: T[key] | undefined }
  top: number
  skip: number
}

/**
 * Returns the possible OData params from a request
 * @param request The incoming request message
 */
export const getOdataParams = <T>(url: string | undefined, entity: Entity<T>) => {
  const order: OrderType<T> = {}

  const parseFilter = (filter: string) => {
    const filterObj: { [key in keyof T]?: T[key] } = {}
    const [fieldName, value] = filter.split(' eq ')
    let parsedValue: any = value.replace(/'/g, '')

    switch (entity.properties.find(f => f.property === fieldName)?.type) {
      case EdmType.Int16:
      case EdmType.Int32:
        parsedValue = parseInt(parsedValue, 10)
        break
      case EdmType.Double:
        parsedValue = parseFloat(parsedValue)
        break
      default:
        break
    }

    filterObj[fieldName as keyof T] = parsedValue as any
    return filterObj
  }

  const getOrderByValue = (value: string): OrderType<T> => {
    value = value.trim()
    if (value.indexOf(',') !== -1) {
      const other = Object.assign({}, ...value.split(',').map(v => getOrderByValue(v)))
      return other
    }
    if (value.indexOf(' ') !== -1) {
      const [field, direction] = value.split(' ')
      order[field as keyof T] = direction.toUpperCase() === 'DESC' ? 'DESC' : 'ASC'
      return order
    }
    order[value as keyof T] = 'ASC'
    return order
  }

  const params = (url && parse(url, true).query) || {}
  const orderBy = (params.$orderby && getOrderByValue(params.$orderby.toString())) || undefined

  const value: OdataParams<T> = {
    ...(params.$skip ? { skip: parseInt(params.$skip as string, 10) } : { skip: 0 }),
    ...(params.$top ? { top: parseInt(params.$top as string, 10) } : { top: 0 }),
    ...(params.$filter ? { filter: parseFilter(params.$filter as string) } : { filter: {} }),
    select:
      (params.$select && ((params.$select as string).split(',') as Array<keyof T>)) ||
      entity.properties.map(p => p.property) ||
      [],
    expand: (params.$expand && ((params.$expand as string).split('(')[0].split(',') as Array<keyof T>)) || [],
    orderBy,
    expandExpression: params.$expand as string,
  }

  return value
}
