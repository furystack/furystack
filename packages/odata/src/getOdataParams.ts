import { parse } from 'url'
import { Entity } from './models'

/**
 * Returns the possible OData params from a request
 * @param request The incoming request message
 */
export const getOdataParams = <T>(url: string | undefined, entity: Entity<T>) => {
  type OrderType = { [key in keyof T]?: 'ASC' | 'DESC' }

  const order = {} as OrderType

  const parseFilter = (filter: string) => {
    const filterObj = {} as { [key in keyof T]?: T[key] }
    const [fieldName, value] = filter.split(' eq ')
    const parsedValue = value.replace(/'/g, '')
    filterObj[fieldName as keyof T] = parsedValue as any
    return filterObj
  }

  const getOrderByValue = (value: string): OrderType => {
    value = value.trim()
    if (value.indexOf(',') !== -1) {
      const other = Object.assign({}, ...value.split(',').map(v => getOrderByValue(v)))
      return other
    }
    if (value.indexOf(' ') !== -1) {
      const [field, direction] = value.split(' ')
      order[field as keyof T] = direction.toUpperCase() === 'DESC' ? 'DESC' : 'ASC'
      return order
    }
    order[value as keyof T] = 'ASC'
    return order
  }

  const params = (url && parse(url, true).query) || {}
  const orderBy = (params.$orderby && getOrderByValue(params.$orderby.toString())) || undefined

  return {
    ...(params.$skip ? { skip: parseInt(params.$skip as string, 10) } : {}),
    ...(params.$top ? { top: parseInt(params.$top as string, 10) } : {}),
    ...(params.$filter ? { filter: parseFilter(params.$filter as string) } : {}),
    select:
      (params.$select && ((params.$select as string).split(',') as Array<keyof T>)) ||
      entity.properties.map(p => p.property),
    expand: (params.$expand && ((params.$expand as string).split('(')[0].split(',') as Array<keyof T>)) || [],
    orderBy,
    expandExpression: params.$expand as string,
  }
}
