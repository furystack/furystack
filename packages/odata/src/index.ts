export * from './InjectorExtension'
export * from './model-builder'
export * from './namespace-builder'
export * from './models'
export * from './xml-utils'
import './InjectorExtension'
