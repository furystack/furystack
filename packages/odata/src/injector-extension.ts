import { HttpApiSettings } from '@furystack/http-api'
import { Injector } from '@furystack/inject'
import { DataSet } from '@furystack/repository'
import { ModelBuilder } from './model-builder'
import { createOdataRouter } from './routing'
import { OdataContext } from './odata-context'

declare module '@furystack/inject/dist/injector' {
  /**
   * Defines an extended Injector instance
   */
  interface Injector {
    /**
     * registers an OData endpoint to the injector.
     * Usage example:
     * ````ts
     * myInjector
     *  .useOdata('odata', builder =>
     *    builder.addNamespace('myNamespace', namespaceBuilder => ... ))
     * ````
     *
     */
    useOdata: (route: string, buildModel: (builder: ModelBuilder) => ModelBuilder) => this
    getOdataContext: <T>() => OdataContext<T> & {
      getCurrentDataSet: () => DataSet<T>
      getCurrentEntity: () => Promise<T | undefined>
    }
  }
}

Injector.prototype.useOdata = function(route, buildModel) {
  const instance = buildModel(new ModelBuilder())
  this.setExplicitInstance(instance)

  const entities = Array.from(instance.namespaces.values())
    .map(ns => ns.entities.entities.values())
    .map(e => Array.from(e))
    .flat()

  const collections = Array.from(instance.namespaces.values())
    .map(ns => ns.collections.collections.values())
    .map(c => Array.from(c))
    .flat()

  const globalActions = [...Array.from(instance.namespaces.values()).map(ns => ns.actions)].flat()

  const globalFunctions = [...Array.from(instance.namespaces.values()).map(ns => ns.functions)].flat()

  this.getInstance(HttpApiSettings).actions.push(
    createOdataRouter({
      route,
      entities,
      collections,
      globalActions,
      globalFunctions,
    }),
  )
  return this
}

Injector.prototype.getOdataContext = function<T>() {
  const context = this.getInstance(OdataContext) as OdataContext<T>
  return {
    ...context,
    getCurrentDataSet: () => this.getDataSetFor<T>(context.collection.name) as DataSet<T>,
    getCurrentEntity: () => this.getDataSetFor<T>(context.collection.name).get(this, context.entityId as any),
  }
}
