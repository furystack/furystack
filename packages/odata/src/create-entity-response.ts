import { Injector } from '@furystack/inject'
import { Repository } from '@furystack/repository'
import { PathHelper } from '@furystack/utils'
import { getOdataParams, OdataParams } from './get-odata-params'
import { Entity } from './models'
import { OdataContext } from './odata-context'

/**
 * Removes the most outer level from the expand expression
 *
 * @param expand The Expand expression
 * @returns the popped expression
 */
export const popExpandLevel = (expand: string) => {
  const initial = expand.split('(')
  initial.shift()
  const withoutPrefix = initial.join('(').split(')')
  withoutPrefix.pop()
  const withoutPostfix = withoutPrefix.join(')')

  const withFirstSegment = withoutPostfix.split('(')
  withFirstSegment[0] = withFirstSegment[0].replace(/;/g, '&')
  const final = withFirstSegment.join('(')
  return final
}

/**
 * Method that adds expanded fields to an entity model
 *
 * @param options The options to be provided
 */
export const createEntityResponse = async <T>(options: {
  injector: Injector
  entity: T
  entityType: Entity<T>
  entityTypes: Array<Entity<any>>
  odataParams: OdataParams<T>
  repo: Repository
  odataContext: OdataContext<T>
}) => {
  const returnEntity: any = {}

  for (const selectField of options.odataParams.select) {
    returnEntity[selectField] = options.entity[selectField]
  }

  const entityId = options.entity[options.entityType.primaryKey]

  returnEntity['@odata.id'] = PathHelper.joinPaths(
    options.odataContext.server,
    options.odataContext.odataRoute,
    (options.odataContext.navigationProperty && options.odataContext.navigationProperty.dataSet) ||
      options.odataContext.collection.name,
    typeof entityId === 'number' ? `(${entityId.toString()})` : `('${entityId}')`,
  )

  if (options.odataParams.expand.length === 0) {
    return returnEntity
  }

  const navExpressions = options.odataParams.expand.map(fieldName => ({
    expandExpression: fieldName,
    navProperty:
      options.entityType.navigationProperties &&
      options.entityType.navigationProperties.find(np => np.propertyName === (fieldName as string).split('(')[0]),
    navPropertyCollection:
      options.entityType.navigationPropertyCollections &&
      options.entityType.navigationPropertyCollections.find(
        np => np.propertyName === (fieldName as string).split('(')[0],
      ),
  }))

  if (!navExpressions.length) {
    return returnEntity
  }

  const expandedEntity = { ...returnEntity }
  await Promise.all(
    navExpressions.map(async navExpression => {
      const { navProperty } = navExpression
      const { navPropertyCollection } = navExpression

      if (navProperty) {
        const dataSet = options.repo.getDataSetFor(navProperty.dataSet || navProperty.relatedModel)
        const entityType = options.entityTypes.find(t => t.model === navProperty.relatedModel) as Entity<
          typeof navProperty.relatedModel
        >

        const poppedExpandLevel = popExpandLevel(options.odataParams.expandExpression)
        const expandOdataParams = options.odataParams.expandExpression
          ? {
              expandExpression: poppedExpandLevel,
              ...getOdataParams(`?${poppedExpandLevel}`, entityType),
            }
          : {}
        const expanded = await navProperty.getRelatedEntity(options.entity, dataSet, options.injector, {
          ...expandOdataParams,
        })
        expandedEntity[navProperty.propertyName] = await createEntityResponse<any>({
          entity: expanded,
          injector: options.injector,
          entityType,
          entityTypes: options.entityTypes,
          odataParams: {
            ...expandOdataParams,
          } as any,
          repo: options.repo,
          odataContext: {
            ...options.odataContext,
            navigationProperty: navProperty,
          },
        })
      } else if (navPropertyCollection) {
        const dataSet = options.repo.getDataSetFor(navPropertyCollection.dataSet || navPropertyCollection.relatedModel)
        const entityType = options.entityTypes.find(t => t.model === navPropertyCollection.relatedModel) as Entity<
          typeof navPropertyCollection.relatedModel
        >
        const poppedExpandLevel = popExpandLevel(options.odataParams.expandExpression)

        const expandedEntities = await Promise.all(
          (
            await navPropertyCollection.getRelatedEntities<any>(
              options.entity,
              dataSet,
              options.injector,
              options.odataParams,
            )
          ).map(async expanded => {
            return await createEntityResponse<any>({
              entity: expanded,
              injector: options.injector,
              entityType,
              entityTypes: options.entityTypes,
              odataParams: {
                ...options.odataParams,
                ...(options.odataParams.expandExpression
                  ? {
                      expandExpression: poppedExpandLevel,
                      expand: getOdataParams(`?${poppedExpandLevel}`, entityType).expand,
                    }
                  : {}),
              },
              repo: options.repo,
              odataContext: options.odataContext,
            })
          }),
        )
        expandedEntity[navPropertyCollection.propertyName] = expandedEntities
      }
    }),
  )
  return expandedEntity
}
