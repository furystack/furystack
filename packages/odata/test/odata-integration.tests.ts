import { Injector } from '@furystack/inject'
import { InMemoryStore } from '@furystack/core'
import got from 'got'
import { JsonResult } from '@furystack/http-api'
import { EdmType } from '../src/index'

describe('OData Integration Tests', () => {
  class User {
    id!: number
    age!: number
    height?: number
    name!: string
  }

  const userList = [
    { id: 1, age: 123, name: 'asdf', height: 0.3 },
    { id: 2, age: 123, name: 'sdf', height: 0.3 },
    { id: 3, age: 456, name: 'asdf', height: 0.2 },
    { id: 4, age: 678, name: 'o8ugzsdgf' },
  ]

  class Group {
    id!: number
    name!: string
    memberIds!: number[]
    ownerId!: number
  }

  const groupList: Group[] = [
    { id: 1, name: 'all users', memberIds: [1, 2, 3, 4], ownerId: 4 },
    { id: 2, name: 'admins', memberIds: [1], ownerId: 1 },
  ]

  class Entity2 {
    guid!: string
    booleanValue?: boolean
    complexValue!: { foo: string; bar: string }
    entity1Id!: number
  }

  const testEntity2List = [
    { guid: 'asdf', entity1Id: 1, booleanValue: true, complexValue: { foo: '', bar: '' } },
    {
      guid: 'sdfg',
      entity1Id: 2,
      booleanValue: true,
      complexValue: { foo: 'f', bar: 'b' },
    },
    { guid: 'fghj', entity1Id: 3, booleanValue: false, complexValue: { foo: 'bar', bar: 'foo' } },
    { guid: 'hjkl', entity1Id: 4, complexValue: { foo: 'asdf', bar: 'asdf' } },
  ]

  const i = new Injector()
  beforeAll(async () => {
    i.setupStores(sm =>
      sm
        .addStore(new InMemoryStore({ model: User, primaryKey: 'id' }))
        .addStore(new InMemoryStore({ model: Group, primaryKey: 'id' }))
        .addStore(new InMemoryStore({ model: Entity2, primaryKey: 'guid' })),
    )
      .setupRepository(repo =>
        repo
          .createDataSet(User, { name: 'UserCollection' })
          .createDataSet(Group, { name: 'GroupCollection' })
          .createDataSet(Entity2, { name: 'Entity2Collection' }),
      )
      .useOdata('odata', odata =>
        odata.addNameSpace('default', ns => {
          ns.setupEntities(entities =>
            entities
              .addEntityType({
                model: User,
                primaryKey: 'id',
                properties: [
                  { property: 'id', type: EdmType.Int16, nullable: false },
                  { property: 'age', type: EdmType.Int16, nullable: false },
                  { property: 'name', type: EdmType.String },
                  { property: 'height', type: EdmType.Double },
                ],
                navigationProperties: [
                  {
                    propertyName: 'MyEntity2',
                    dataSet: 'Entity2Collection',
                    relatedModel: Entity2,
                    getRelatedEntity: async (entity, dataset, injector, filter) => {
                      const related = await dataset.filter(injector, {
                        ...filter,
                        filter: { ...filter.filter, entity1Id: entity.id },
                        top: 1,
                      })
                      return related[0] || null
                    },
                  },
                ],
              })
              .addEntityType({
                model: Group,
                primaryKey: 'id',
                properties: [
                  { property: 'id', type: EdmType.Int16, nullable: false },
                  { property: 'name', nullable: false, type: EdmType.String },
                ],
                navigationPropertyCollections: [
                  {
                    dataSet: 'GroupCollection',
                    relatedModel: Group,
                    propertyName: 'memberIds',
                    getRelatedEntities: async (entity, dataSet, injector, filter) => {
                      const members = await dataSet.filter(injector, {
                        ...filter,
                        filter: {
                          ...filter,
                          id: { $in: entity.memberIds },
                        },
                      })
                      return [...members]
                    },
                  },
                ],
              })
              .addEntityType({
                model: Entity2,
                primaryKey: 'guid',
                properties: [
                  {
                    property: 'guid',
                    type: EdmType.Guid,
                    nullable: false,
                  },
                  {
                    property: 'booleanValue',
                    type: EdmType.Boolean,
                  },
                  {
                    property: 'complexValue',
                    type: EdmType.Unknown,
                    nullable: false,
                  },
                ],
              }),
          )
          ns.setupCollections(collections =>
            collections
              .addCollection({
                model: User,
                name: 'UserCollection',
                actions: [
                  { name: 'entity1CustomAction', action: async () => JsonResult({ result: 'CustomActionResult' }) },
                ],
                functions: [
                  { name: 'entity1CustomFunction', action: async () => JsonResult({ result: 'CustomFunctionResult' }) },
                ],
              })
              .addCollection({
                model: Entity2,
                name: 'Entity2Collection',
                actions: [
                  { name: 'entity2CustomAction', action: async () => JsonResult({ result: 'CustomActionResult' }) },
                ],
                functions: [
                  { name: 'entity2CustomFunction', action: async () => JsonResult({ result: 'CustomFunctionResult' }) },
                ],
              }),
          )
          ns.setupGlobalActions([
            { name: 'globalAction1', action: async () => JsonResult({ result: 'CustomActionResult' }) },
          ])
          ns.setupGlobalFunctions([
            { name: 'globalFunction1', action: async () => JsonResult({ result: 'CustomFunctionResult' }) },
          ])
          return ns
        }),
      )
      .listenHttp({ port: 8888 })

    const entity1store = i.getDataSetFor<User>('UserCollection')

    for (const testEntity1 of userList) {
      await entity1store.add(i, testEntity1)
    }

    const entity2store = i.getDataSetFor<Entity2>('Entity2Collection')

    for (const tetstEntiy2 of testEntity2List) {
      await entity2store.add(i, tetstEntiy2)
    }
  })

  describe('Metadata', () => {
    it('Root should match the snapshot', async () => {
      const rootResponse = await got(`http://localhost:8888/odata`)
      expect(JSON.parse(rootResponse.body)).toMatchSnapshot()
    })

    it('Should have metadata', async () => {
      const metadataXML = await got(`http://localhost:8888/odata/$metadata`)
      expect(metadataXML.body).toMatchSnapshot()
    })
  })

  describe('Collection tests', () => {
    it('Retrieve a plain collection', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection`)
      const body = JSON.parse(response.body)
      expect(body['@odata.context']).toBe('odata/$metadata#UserCollection')
      expect(body['@odata.count']).toBe(4)
      expect(body.value.length).toBe(4)
      for (const testEntity of userList) {
        const test = { ...testEntity, '@odata.id': `http://localhost:8888/odata/UserCollection/(${testEntity.id})` }
        expect(body.value).toContainEqual(test)
      }
    })

    it('Retrieve from collection with a $select=id,name filter', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$select=id,name`)
      const body = JSON.parse(response.body)
      expect(body['@odata.context']).toBe('odata/$metadata#UserCollection')
      expect(body['@odata.count']).toBe(4)
      expect(body.value.length).toBe(4)
      for (const testEntity of userList) {
        const test = {
          id: testEntity.id,
          name: testEntity.name,
          '@odata.id': `http://localhost:8888/odata/UserCollection/(${testEntity.id})`,
        }
        expect(body.value).toContainEqual(test)
      }
    })

    it('Retrieve from collection with a $top=2 filter', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$top=2`)
      const body = JSON.parse(response.body)
      expect(body.value.length).toEqual(2)
      expect(body.value.map((v: User) => v.id)).toEqual([1, 2])
    })

    it('Retrieve from collection with a $skip=2 filter', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$skip=2`)
      const body = JSON.parse(response.body)
      expect(body.value.length).toEqual(2)
      expect(body.value.map((v: User) => v.id)).toEqual([3, 4])
    })
    it('Retrieve from collection with a "$filter=name eq asdf" filter on strings', async () => {
      const response = await got(
        `http://localhost:8888/odata/UserCollection?$filter=${encodeURIComponent('name eq asdf')}`,
      )
      const body = JSON.parse(response.body)
      for (const value of body.value.map((v: User) => v.name)) {
        expect(value).toBe('asdf')
      }
    })

    it('Retrieve from collection with a "$filter=age eq 123" filter (ints)', async () => {
      const response = await got(
        `http://localhost:8888/odata/UserCollection?$filter=${encodeURIComponent('age eq 123')}`,
      )
      const body = JSON.parse(response.body)
      for (const value of body.value.map((v: User) => v.age)) {
        expect(value).toBe(123)
      }
    })

    it('Retrieve from collection with "$filter=height eq 0.3" filter (floats)', async () => {
      const response = await got(
        `http://localhost:8888/odata/UserCollection?$filter=${encodeURIComponent('height eq 0.3')}`,
      )
      const body = JSON.parse(response.body)
      for (const value of body.value.map((v: User) => v.height)) {
        expect(value).toEqual(0.3)
      }
    })

    it('Retrieve from collection with a $orderby param (fall back to ASC)', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$orderby=height`)
      const body = JSON.parse(response.body)
      expect(body.value.map((v: User) => v.height)).toEqual([0.2, 0.3, 0.3, undefined])
    })

    it('Retrieve from collection with a $orderby ASC param', async () => {
      const response = await got(
        `http://localhost:8888/odata/UserCollection?$orderby=${encodeURIComponent('height ASC')}`,
      )
      const body = JSON.parse(response.body)
      expect(body.value.map((v: User) => v.height)).toEqual([0.2, 0.3, 0.3, undefined])
    })

    it('Retrieve entities from collection with a $orderby DESC param', async () => {
      const response = await got(
        `http://localhost:8888/odata/UserCollection?$orderby=${encodeURIComponent('height DESC')}`,
      )
      const body = JSON.parse(response.body)
      expect(body.value.map((v: User) => v.height)).toEqual([0.3, 0.3, 0.2, undefined])
    })

    it('Retrieve entities from collection (Entity2)', async () => {
      const response = await got(`http://localhost:8888/odata/Entity2Collection/`)
      const body = JSON.parse(response.body)
      expect(body['@odata.context']).toBe('odata/$metadata#Entity2Collection')
      expect(body['@odata.count']).toBe(4)
      expect(body.value.length).toBe(4)
      for (const testEntity of testEntity2List) {
        const test = {
          guid: testEntity.guid,
          booleanValue: testEntity.booleanValue,
          complexValue: testEntity.complexValue,
          '@odata.id': `http://localhost:8888/odata/Entity2Collection/('${testEntity.guid}')`,
        }
        expect(body.value).toContainEqual(test)
      }
    })
  })

  describe('Navigation properties', () => {
    it('Retrieve entities from collection (Entity1) with a $expand=MyEntity2', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$expand=MyEntity2`)
      const body = JSON.parse(response.body)
      expect(body['@odata.context']).toBe('odata/$metadata#UserCollection')
      expect(body['@odata.count']).toBe(4)
      expect(body.value.length).toBe(4)
      for (const testEntity of userList) {
        const { ...relatedT2 } = testEntity2List.find(t2 => t2.entity1Id === testEntity.id)
        delete relatedT2.entity1Id
        const test = {
          ...testEntity,
          MyEntity2: {
            ...relatedT2,
            '@odata.id': `http://localhost:8888/odata/Entity2Collection/('${relatedT2?.guid}')`,
          },
          '@odata.id': `http://localhost:8888/odata/UserCollection/(${testEntity.id})`,
        }
        expect(body.value).toContainEqual(test)
      }
    })

    it('Retrieve entities from collection (Entity1) with a $expand=MyEntity2&select=id,MyEntity2', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection?$expand=MyEntity2&$select=id,MyEntity2`)
      const body = JSON.parse(response.body)
      expect(body['@odata.context']).toBe('odata/$metadata#UserCollection')
      expect(body['@odata.count']).toBe(4)
      expect(body.value.length).toBe(4)
      for (const testEntity of userList) {
        const { ...relatedT2 } = testEntity2List.find(t2 => t2.entity1Id === testEntity.id)
        delete relatedT2.entity1Id
        const test = {
          id: testEntity.id,
          MyEntity2: {
            ...relatedT2,
            '@odata.id': `http://localhost:8888/odata/Entity2Collection/('${relatedT2?.guid}')`,
          },
          '@odata.id': `http://localhost:8888/odata/UserCollection/(${testEntity.id})`,
        }
        expect(body.value).toContainEqual(test)
      }
    })

    it('Retrieve referenced entity through a navigation path ', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection(1)/MyEntity2`)
      const body = JSON.parse(response.body)
      const entity = testEntity2List[0]
      expect(body).toStrictEqual({
        '@odata.context': '$metadata#default.Entity2',
        '@odata.id': `http://localhost:8888/odata/Entity2Collection/('${entity.guid}')`,
        guid: entity.guid,
        complexValue: entity.complexValue,
        booleanValue: entity.booleanValue,
      })
    })

    it('Retrieve referenced entity through a navigation path and $select ', async () => {
      const response = await got(`http://localhost:8888/odata/UserCollection(1)/MyEntity2?$select=guid,booleanValue`)
      const body = JSON.parse(response.body)
      const entity = testEntity2List[0]
      expect(body).toStrictEqual({
        '@odata.context': '$metadata#default.Entity2',
        '@odata.id': `http://localhost:8888/odata/Entity2Collection/('${entity.guid}')`,
        guid: entity.guid,
        booleanValue: entity.booleanValue,
      })
    })
  })

  afterAll(async () => {
    await i.dispose()
  })
})
