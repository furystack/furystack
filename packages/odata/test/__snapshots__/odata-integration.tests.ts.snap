// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OData Integration Tests Metadata Root should match the snapshot 1`] = `
Object {
  "@odata.context": "http://localhost:8888/odata/$metadata",
  "value": Array [
    Object {
      "kind": "EntitySet",
      "name": "UserCollection",
      "url": "UserCollection",
    },
    Object {
      "kind": "EntitySet",
      "name": "Entity2Collection",
      "url": "Entity2Collection",
    },
  ],
}
`;

exports[`OData Integration Tests Metadata Should have metadata 1`] = `"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><edmx:Edmx Version=\\"4.0\\" xmlns:edmx=\\"http://docs.oasis-open.org/odata/ns/edmx\\"><edmx:DataServices ><Schema xmlns=\\"http://docs.oasis-open.org/odata/ns/edm\\" Namespace=\\"default\\"><EntityType Name=\\"User\\"><Key ><PropertyRef Name=\\"id\\" /></Key><Property Name=\\"id\\" Type=\\"Edm.Int16\\" Nullable=\\"false\\" /><Property Name=\\"age\\" Type=\\"Edm.Int16\\" Nullable=\\"false\\" /><Property Name=\\"name\\" Type=\\"Edm.String\\" /><Property Name=\\"height\\" Type=\\"Edm.Double\\" /><NavigationProperty Name=\\"MyEntity2\\" Type=\\"default.Entity2\\" /></EntityType><EntityType Name=\\"Group\\"><Key ><PropertyRef Name=\\"id\\" /></Key><Property Name=\\"id\\" Type=\\"Edm.Int16\\" Nullable=\\"false\\" /><Property Name=\\"name\\" Type=\\"Edm.String\\" Nullable=\\"false\\" /><NavigationProperty Name=\\"memberIds\\" Type=\\"Collection(default.Group)\\" /></EntityType><EntityType Name=\\"Entity2\\"><Key ><PropertyRef Name=\\"guid\\" /></Key><Property Name=\\"guid\\" Type=\\"Edm.Guid\\" Nullable=\\"false\\" /><Property Name=\\"booleanValue\\" Type=\\"Edm.Boolean\\" /><Property Name=\\"complexValue\\" Type=\\"Edm.Unknown\\" Nullable=\\"false\\" /></EntityType><EntityContainer Name=\\"DefaultContainer\\"><EntitySet Name=\\"UserCollection\\" EntityType=\\"default.User\\" /><EntitySet Name=\\"Entity2Collection\\" EntityType=\\"default.Entity2\\" /></EntityContainer><Action Name=\\"globalAction1\\"></Action><Function Name=\\"globalFunction1\\"></Function><Action Name=\\"entity1CustomAction\\" IsBound=\\"true\\"><Parameter Name=\\"bindingParameter\\" Type=\\"Collection(default.User)\\" /></Action><Function Name=\\"entity1CustomFunction\\" IsBound=\\"true\\"><Parameter Name=\\"bindingParameter\\" Type=\\"Collection(default.User)\\" /></Function><Action Name=\\"entity2CustomAction\\" IsBound=\\"true\\"><Parameter Name=\\"bindingParameter\\" Type=\\"Collection(default.Entity2)\\" /></Action><Function Name=\\"entity2CustomFunction\\" IsBound=\\"true\\"><Parameter Name=\\"bindingParameter\\" Type=\\"Collection(default.Entity2)\\" /></Function></Schema></edmx:DataServices></edmx:Edmx>"`;
